patches/plus_one.patch

```diff
@@ -0,0 +1,777 @@
++    uint32_t attempts = 0;
++    uint32_t index = serverId->indexNumber();
++    
++    for (attempts = 0; attempts < totalAttempts; attempts++) {
++        applyTrackerChanges();
++        index++;
++        if (index > tracker.size()) {
++            index = 1;
++        }
++        ServerId id = tracker.getServerIdAtIndexWithService(
++            index, WireFormat::BACKUP_SERVICE);
++        if (id.isValid() &&
++            !conflictWithAny(id, numBackups, backupIds)) {
++            okToLogNextProblem = true;
++            return id;
++        }
++    }
++    if (okToLogNextProblem) {
++        RAMCLOUD_CLOG(WARNING, "PlusOneBackupSelector could not find a "
++            "suitable server in %d attempts; may need to wait for additional "
++            "servers to enlist",
++            attempts);
++        okToLogNextProblem = false;
++    }
++    return ServerId(/* Invalid */);
++}
++
++} // namespace RAMCloud
+diff --git a/src/PlusOneBackupSelector.h b/src/PlusOneBackupSelector.h
+new file mode 100644
+index 00000000..794a9285
+--- /dev/null
++++ b/src/PlusOneBackupSelector.h
+@@ -0,0 +1,41 @@
++/* Copyright (c) 2011-2019 Stanford University
++ *
++ * Permission to use, copy, modify, and distribute this software for any
++ * purpose with or without fee is hereby granted, provided that the above
++ * copyright notice and this permission notice appear in all copies.
++ *
++ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR(S) DISCLAIM ALL WARRANTIES
++ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
++ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL AUTHORS BE LIABLE FOR
++ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
++ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
++ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
++ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
++ */
++
++#ifndef RAMCLOUD_PLUSONEBACKUPSELECTOR_H
++#define RAMCLOUD_PLUSONEBACKUPSELECTOR_H
++
++#include "Common.h"
++#include "BackupSelector.h"
++
++namespace RAMCloud {
++
++/**
++ * Selects backups to store replicas starting with (masterServerId+1)%n
++ */
++class PlusOneBackupSelector : public BackupSelector {
++  PUBLIC:
++    explicit PlusOneBackupSelector(Context* context,
++                                       const ServerId* serverId,
++                                       uint32_t numReplicas,
++                                       bool allowLocalBackup);
++    ServerId selectSecondary(uint32_t numBackups, const ServerId backupIds[]);
++
++  PRIVATE:
++    DISALLOW_COPY_AND_ASSIGN(PlusOneBackupSelector);
++};
++
++} // namespace RAMCloud
++
++#endif
+diff --git a/src/PlusOneBackupSelectorTest.cc b/src/PlusOneBackupSelectorTest.cc
+new file mode 100644
+index 00000000..ab0485ab
+--- /dev/null
++++ b/src/PlusOneBackupSelectorTest.cc
+@@ -0,0 +1,101 @@
++/* Copyright (c) 2009-2016 Stanford University
```

Comment:
2019 is the current year. I have no idea how copyright statements work though. I don't know if we should just omit this entirely on files that we create? :man_shrugging: